### ISSUE TITLE
Support custom json(b) parser

### ISSUE THREAD

**Issue by qnikst** [nikita-volkov/hasql-th#3]
I have following pattern that I use a lot on my code:

```
(HD.column.HD.nonNullable) (HD.jsonbBytes (first T.pack . eitherDecodeStrict)
```

This way I can parse json structure into desired type, and get a `RowError` in case of error.
Now if I use `json` or `jsonb` l get `Value` out. And I have to parse it in my application code, and introduce additional errors, it makes code quite complex.

I wonder if there is any other pattern that can be used to achieve the same goal.

Maybe there is a way that we can use to extend parser, for example have a function that will take a configuration and will build quasi-quoter. In configuration it's possible to pass type name and corresponding encoders and decoders, then it can allow parser extending without affecting the library.

**Comment by nikita-volkov** [comment]
I've been thinking about introducing some sort of a failing version of `fmap` for decoder. It seems like with the advent of this library `Statement` also needs it.

I'm talking about something like this:

```haskell
mapWithErr :: (a -> Either Text b) -> Statement params a -> Statement params b
```

Would that suffice?

**Comment by qnikst** [comment]
It will definitely solve the issue as I will be able to do some post processing before creating my type.  And it looks like a nice function to have.
However I'm not sure if I'm happy with that solution for exactly my problem. With such function I loose "locality" of the check, previously I was able to check the field encoding in a place where I'm decoding the field, now I can check entire structure only. It means that from Statement I need to return some type `a` to do the check and it seems I'll have to return a tuple, and then write a function that will convert a tuple into my business type with some additional checks so I'm afraid I may loose the benefits that hasql-th provides and appear with similar amount of code, that may be even harder to read.

```Haskell
foo = dimap coerce coerce $ reify (V.mapM check) [vectorStatement| select n::int8, fields::jsonb, params::jsonb where contest_id=$1|] where
  check  :: Vector (Int64, Value, Value) -> Either Text (Vector MyType)
  check (n, v1, v2) =
    case MyType <$> pure n <*> fromJSON v1 <*> fromJSON v2 of
      Success r -> Right r
      Error s -> Left $ T.pack s
```

This thing looks hard to read, and the things get worse if I have wide structure with many fields.

Compare that to my current code:

```Haskell
foo = HS.Statement sql encoder decoder where
  sql = [hasql|select n::int8, fields::jsonb, params::jsonb where contest_id=$1|]
  encoder = contramap coerce ((HE.param.HE.nonNullable) HE.int8)
  decoder = HD.rowVector $ do
     typeN <- (HD.column.HD.nonNullable) (HD.int8) <&> coerced
     typeField <- (HD.column.HD.nonNullable) (HD.jsonbBytes (first T.pack . eitherDecodeStrict))
     typeParams <- (HD.column.HD.nonNullable) (HD.jsonbBytes (first T.pack . eitherDecodeStrict))
     pure MyType{..}
```

On the one hand there are more lines, but they are well structured and easy to read, and it's very handy to change them or add additional ones, it scales to more columns quite well. 

And here is ideal solution for me:

1. Have a way to define additional  types (not sure about the API)

```Haskell
myVectorStatement = mkVectorStatement defaultParams
  { extraDecoders =
      [("aesonb", [| HD.jsonbBytes (first T.pack . eitherDecodeStrict)) |] )] 
  }
```

2. Use that decoder:

```Haskell
foo = dimap coerce coerce
  [myVectorStatement| select n::int8, fields::aejsonb, params::aesonb where contest_id=$1|] 
```

Here all checks are local, I don't need to introduce temporary structures, I can control all my additional types in a single place.

**Comment by k0001** [comment]
> ```haskell
> refine :: (a -> Either Text b) -> Statement params a -> Statement params b
> ```

I agree that something like `refine`,  or generally speaking, something that can achieve more or less what one can accomplish with `hasql`'s [`custom`](https://hackage.haskell.org/package/hasql-1.4.0.1/docs/Hasql-Decoders.html#v:custom) is necessary in `hasql-th`. For what is worth, this is why I haven't been able to switch from `hasql` to `hasql-th` yet.

**Comment by nikita-volkov** [comment]
Hasql 1.4.2 provides the following function in the `Hasql.Statement` module:

```haskell
refineResult :: (a -> Either Text b) -> Statement params a -> Statement params b
```

Can we consider this issue resolved?

**Comment by qnikst** [comment]
For me it has not that nice ergonomics but it allows to solve the problem. So we can consider this issue resolved.
